Add asyncio stack debugging to USR1 signal

konradkonrad opened the pull request:
This adds `print_stack()` calls for all current (if any) `asyncio` tasks
to the already existing debugging signal hook listening for `SIGUSR1`.

For reference: https://github.com/raiden-network/raiden/issues/7141#issuecomment-901757801

konradkonrad committed:
Add asyncio stack debugging to USR1 signal

This adds `print_stack()` calls for all current (if any) `asyncio` tasks
to the already existing debugging signal hook listening for `SIGUSR1`.

For reference: https://github.com/raiden-network/raiden/issues/7141#issuecomment-901757801

konradkonrad commented:
This now creates output like the below:
```
pytest -s raiden/tests/unit/network/rtc/test_web_rtc.py # add some longer sleep in the test, because it is hard to catch otherwise
# and run
kill -SIGUSR1 $(pidof -x pytest)
```

```
********************************************************************************
* Threads
********************************************************************************
Thread 0x7f278b7fe640 (idle threadpool worker)

********************************************************************************
Thread 0x7f278bfff640 (idle threadpool worker)

********************************************************************************
Thread 0x7f27a1034640 (idle threadpool worker)

********************************************************************************
Thread 0x7f27a1835640 (idle threadpool worker)

********************************************************************************
Thread 0x7f27a2036640 (asyncio_3)

  File "/usr/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.9/concurrent/futures/thread.py", line 75, in _worker
    work_item = work_queue.get(block=True)
  File "/usr/lib/python3.9/queue.py", line 294, in get
    if not self._count.acquire(block, timeout):
  File "/usr/lib/python3.9/threading.py", line 443, in acquire
    self._cond.wait(timeout)
  File "/usr/lib/python3.9/threading.py", line 312, in wait
    waiter.acquire()

********************************************************************************
Thread 0x7f27a2837640 (asyncio_2)

  File "/usr/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.9/concurrent/futures/thread.py", line 75, in _worker
    work_item = work_queue.get(block=True)
  File "/usr/lib/python3.9/queue.py", line 294, in get
    if not self._count.acquire(block, timeout):
  File "/usr/lib/python3.9/threading.py", line 443, in acquire
    self._cond.wait(timeout)
  File "/usr/lib/python3.9/threading.py", line 312, in wait
    waiter.acquire()

********************************************************************************
Thread 0x7f27a3038640 (asyncio_1)

  File "/usr/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.9/concurrent/futures/thread.py", line 75, in _worker
    work_item = work_queue.get(block=True)
  File "/usr/lib/python3.9/queue.py", line 294, in get
    if not self._count.acquire(block, timeout):
  File "/usr/lib/python3.9/threading.py", line 443, in acquire
    self._cond.wait(timeout)
  File "/usr/lib/python3.9/threading.py", line 312, in wait
    waiter.acquire()

********************************************************************************
Thread 0x7f27a3839640 (asyncio_0)

  File "/usr/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.9/concurrent/futures/thread.py", line 75, in _worker
    work_item = work_queue.get(block=True)
  File "/usr/lib/python3.9/queue.py", line 294, in get
    if not self._count.acquire(block, timeout):
  File "/usr/lib/python3.9/threading.py", line 443, in acquire
    self._cond.wait(timeout)
  File "/usr/lib/python3.9/threading.py", line 312, in wait
    waiter.acquire()

********************************************************************************
Thread 0x7f27afa81740 (MainThread) (CURRENT)

  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/hub.py", line 647, in run
    loop.run()
  File "/home/konrad/Projects/raiden/raiden/utils/debugging.py", line 38, in on_signal
    gevent.util.print_run_info()
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 97, in print_run_info
    lines = format_run_info(thread_stacks=thread_stacks,
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 152, in format_run_info
    _format_thread_info(lines, thread_stacks, limit, current_thread_ident)
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 189, in _format_thread_info
    lines.append(''.join(traceback.format_stack(frame, limit)))

********************************************************************************
* Greenlets
********************************************************************************
---- Thread boundary
<greenlet.greenlet object at 0x7f27a9f1a880>
 :    Parent: None
 :    Running:
 :      File "/home/konrad/.virtualenvs/raiden/bin/pytest", line 8, in <module>
 :        sys.exit(console_main())
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185, in console_main
 :        code = main()
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162, in main
 :        ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :        return self._hookexec(self, self.get_hookimpls(), kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :        return self._inner_hookexec(hook, methods, kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :        self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :        res = hook_impl.function(*args)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/main.py", line 316, in pytest_cmdline_main
 :        return wrap_session(config, _main)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/main.py", line 269, in wrap_session
 :        session.exitstatus = doit(config, session) or 0
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/main.py", line 323, in _main
 :        config.hook.pytest_runtestloop(session=session)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :        return self._hookexec(self, self.get_hookimpls(), kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :        return self._inner_hookexec(hook, methods, kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :        self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :        res = hook_impl.function(*args)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/main.py", line 348, in pytest_runtestloop
 :        item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :        return self._hookexec(self, self.get_hookimpls(), kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :        return self._inner_hookexec(hook, methods, kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :        self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :        res = hook_impl.function(*args)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/flaky/flaky_pytest_plugin.py", line 94, in pytest_runtest_protocol
 :        self.runner.pytest_runtest_protocol(item, nextitem)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 109, in pytest_runtest_protocol
 :        runtestprotocol(item, nextitem=nextitem)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 126, in runtestprotocol
 :        reports.append(call_and_report(item, "call", log))
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/flaky/flaky_pytest_plugin.py", line 138, in call_and_report
 :        call = call_runtest_hook(item, when, **kwds)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 254, in call_runtest_hook
 :        return CallInfo.from_call(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 311, in from_call
 :        result: Optional[TResult] = func()
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 255, in <lambda>
 :        lambda: ihook(item=item, **kwds), when=when, reraise=reraise
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :        return self._hookexec(self, self.get_hookimpls(), kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :        return self._inner_hookexec(hook, methods, kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :        self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :        res = hook_impl.function(*args)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 162, in pytest_runtest_call
 :        item.runtest()
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
 :        self.ihook.pytest_pyfunc_call(pyfuncitem=self)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :        return self._hookexec(self, self.get_hookimpls(), kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :        return self._inner_hookexec(hook, methods, kwargs)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :        self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :        res = hook_impl.function(*args)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
 :        result = testfunction(**testargs)
 :      File "/home/konrad/Projects/raiden/raiden/tests/unit/network/rtc/test_web_rtc.py", line 57, in test_basics
 :        gevent.sleep(30.1)
 :      File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/hub.py", line 166, in sleep
 :        hub.wait(t)
 +--- <Greenlet "WebRTCManager|Greenlet-0" at 0x7f27a3a43d00: <bound method Runnable._run of <raiden.network.transport.matrix.rtc.web_rtc.WebRTCManager object at 0x7f27a3be63d0>>>; not running
 :          Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :          Spawned at:
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 162, in pytest_runtest_call
 :              item.runtest()
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
 :              self.ihook.pytest_pyfunc_call(pyfuncitem=self)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :              return self._hookexec(self, self.get_hookimpls(), kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :              return self._inner_hookexec(hook, methods, kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :              self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :              res = hook_impl.function(*args)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
 :              result = testfunction(**testargs)
 :            File "/home/konrad/Projects/raiden/raiden/tests/unit/network/rtc/test_web_rtc.py", line 46, in test_basics
 :              manager1 = WebRTCManager(address1, _ignore_web_rtc_messages, _send_to_2, stop_event1)
 :            File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 469, in __init__
 :              super().__init__()
 :            File "/home/konrad/Projects/raiden/raiden/utils/runnable.py", line 26, in __init__
 :              self.greenlet = Greenlet(self._run, *self.args, **self.kwargs)
 +--- <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :          Parent: <greenlet.greenlet object at 0x7f27a9f1a880>
 :          Running:
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/hub.py", line 647, in run
 :              loop.run()
 :            File "/home/konrad/Projects/raiden/raiden/utils/debugging.py", line 38, in on_signal
 :              gevent.util.print_run_info()
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 97, in print_run_info
 :              lines = format_run_info(thread_stacks=thread_stacks,
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 153, in format_run_info
 :              _format_greenlet_info(lines, greenlet_stacks, limit)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 209, in _format_greenlet_info
 :              lines.extend(tree.format_lines(details={
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 361, in format_lines
 :              for l in self._render(tree)]
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 434, in _render
 :              self.__render_children(tree)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 454, in __render_children
 :              child_tree = child._render(tree.deeper())
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 417, in _render
 :              self.__render_tb(tree, 'Running:', self.greenlet.gr_frame,
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 377, in __render_tb
 :              tb = ''.join(traceback.format_stack(frame, limit))
 +--- <Greenlet "WebRTCManager|Greenlet-1" at 0x7f27a3a43e10: <bound method Runnable._run of <raiden.network.transport.matrix.rtc.web_rtc.WebRTCManager object at 0x7f27a3be64c0>>>; not running
 :          Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :          Spawned at:
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 162, in pytest_runtest_call
 :              item.runtest()
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
 :              self.ihook.pytest_pyfunc_call(pyfuncitem=self)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :              return self._hookexec(self, self.get_hookimpls(), kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :              return self._inner_hookexec(hook, methods, kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :              self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :              res = hook_impl.function(*args)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
 :              result = testfunction(**testargs)
 :            File "/home/konrad/Projects/raiden/raiden/tests/unit/network/rtc/test_web_rtc.py", line 47, in test_basics
 :              manager2 = WebRTCManager(address2, _ignore_web_rtc_messages, _send_to_1, stop_event2)
 :            File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 469, in __init__
 :              super().__init__()
 :            File "/home/konrad/Projects/raiden/raiden/utils/runnable.py", line 26, in __init__
 :              self.greenlet = Greenlet(self._run, *self.args, **self.kwargs)
 +--- <Greenlet "Greenlet-2" at 0x7f27a3a43bf0: <bound method EventLoop.run_forever of <EventLoop running=True closed=False debug=False>>>
 :          Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :          Running:
 :            File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 193, in run_forever
 :              super(EventLoop, self).run_forever()
 :            File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
 :              self._run_once()
 :            File "/usr/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
 :              event_list = self._selector.select(timeout)
 :            File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 153, in select
 :              self._event.wait()
 :          Spawned at:
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/fixtures.py", line 687, in _compute_fixture_value
 :              fixturedef.execute(request=subrequest)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/fixtures.py", line 1072, in execute
 :              result = hook.pytest_fixture_setup(fixturedef=self, request=request)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
 :              return self._hookexec(self, self.get_hookimpls(), kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
 :              return self._inner_hookexec(hook, methods, kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
 :              self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
 :              res = hook_impl.function(*args)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/fixtures.py", line 1126, in pytest_fixture_setup
 :              result = call_fixture_func(fixturefunc, request, kwargs)
 :            File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/fixtures.py", line 925, in call_fixture_func
 :              fixture_result = next(generator)
 :            File "/home/konrad/Projects/raiden/conftest.py", line 534, in asyncio_loop
 :              event_loop = setup_asyncio_event_loop(RuntimeError)
 :            File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/utils.py", line 22, in setup_asyncio_event_loop
 :              gevent.spawn(new_event_loop.run_forever)
 |    +--- <Greenlet "Greenlet-3" at 0x7f27a39b4d00: _run>; finished
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Spawned at:
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 193, in run_forever
 :                    super(EventLoop, self).run_forever()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
 :                    self._run_once()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
 :                    handle._run()
 :                  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
 :                    self._context.run(self._callback, *self._args)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 83, in task_done
 :                    self.schedule_greenlet(callback, result.result(), *args, **kwargs)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 92, in schedule_greenlet
 :                    greenlet = gevent.spawn(func, *args, **kwargs)
 |    +--- <Greenlet "Greenlet-4" at 0x7f27a39b4e10: _run>; finished
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Spawned at:
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 193, in run_forever
 :                    super(EventLoop, self).run_forever()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
 :                    self._run_once()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
 :                    handle._run()
 :                  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
 :                    self._context.run(self._callback, *self._args)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 83, in task_done
 :                    self.schedule_greenlet(callback, result.result(), *args, **kwargs)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 92, in schedule_greenlet
 :                    greenlet = gevent.spawn(func, *args, **kwargs)
 |    +--- <Greenlet "Greenlet-6" at 0x7f27a3852ae0: _run>; finished
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Spawned at:
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py", line 228, in __setState
 :                    self.emit("statechange")
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 108, in emit
 :                    handled = self._call_handlers(event, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 91, in _call_handlers
 :                    self._emit_run(f, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_asyncio.py", line 40, in _emit_run
 :                    coro = f(*args, **kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcpeerconnection.py", line 1193, in __updateIceGatheringState
 :                    self.emit("icegatheringstatechange")
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 108, in emit
 :                    handled = self._call_handlers(event, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 91, in _call_handlers
 :                    self._emit_run(f, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_asyncio.py", line 40, in _emit_run
 :                    coro = f(*args, **kwargs)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 440, in _on_ice_gathering_state_change
 :                    self.schedule_greenlet(self._handle_candidates_callback, candidates=candidates)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 92, in schedule_greenlet
 :                    greenlet = gevent.spawn(func, *args, **kwargs)
 |    +--- <Greenlet "Greenlet-7" at 0x7f27a3852bf0: _run>; finished
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Spawned at:
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py", line 228, in __setState
 :                    self.emit("statechange")
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 108, in emit
 :                    handled = self._call_handlers(event, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 91, in _call_handlers
 :                    self._emit_run(f, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_asyncio.py", line 40, in _emit_run
 :                    coro = f(*args, **kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcpeerconnection.py", line 1193, in __updateIceGatheringState
 :                    self.emit("icegatheringstatechange")
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 108, in emit
 :                    handled = self._call_handlers(event, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_base.py", line 91, in _call_handlers
 :                    self._emit_run(f, args, kwargs)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pyee/_asyncio.py", line 40, in _emit_run
 :                    coro = f(*args, **kwargs)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 440, in _on_ice_gathering_state_change
 :                    self.schedule_greenlet(self._handle_candidates_callback, candidates=candidates)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 92, in schedule_greenlet
 :                    greenlet = gevent.spawn(func, *args, **kwargs)
 |    +--- <Greenlet "Greenlet-8" at 0x7f27a3852d00: _run>; finished
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Spawned at:
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 193, in run_forever
 :                    super(EventLoop, self).run_forever()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
 :                    self._run_once()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
 :                    handle._run()
 :                  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
 :                    self._context.run(self._callback, *self._args)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 83, in task_done
 :                    self.schedule_greenlet(callback, result.result(), *args, **kwargs)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 92, in schedule_greenlet
 :                    greenlet = gevent.spawn(func, *args, **kwargs)
 |    +--- <Greenlet "Greenlet-9" at 0x7f27a3852e10: timeout_cb(<gevent._gevent_cevent.Event object at 0x7f27a383a)>
 :                Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
 :                Running:
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/hub.py", line 647, in run
 :                    loop.run()
 :                  File "/home/konrad/Projects/raiden/raiden/utils/debugging.py", line 38, in on_signal
 :                    gevent.util.print_run_info()
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 97, in print_run_info
 :                    lines = format_run_info(thread_stacks=thread_stacks,
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 153, in format_run_info
 :                    _format_greenlet_info(lines, greenlet_stacks, limit)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 209, in _format_greenlet_info
 :                    lines.extend(tree.format_lines(details={
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 361, in format_lines
 :                    for l in self._render(tree)]
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 434, in _render
 :                    self.__render_children(tree)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 454, in __render_children
 :                    child_tree = child._render(tree.deeper())
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 434, in _render
 :                    self.__render_children(tree)
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 454, in __render_children
 :                    child_tree = child._render(tree.deeper())
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 417, in _render
 :                    self.__render_tb(tree, 'Running:', self.greenlet.gr_frame,
 :                  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/gevent/util.py", line 377, in __render_tb
 :                    tb = ''.join(traceback.format_stack(frame, limit))
 :                Spawned at:
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 193, in run_forever
 :                    super(EventLoop, self).run_forever()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
 :                    self._run_once()
 :                  File "/usr/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
 :                    event_list = self._selector.select(timeout)
 :                  File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/aiogevent.py", line 151, in select
 :                    gevent.spawn_later(timeout, timeout_cb, self._event)
 +--- <Greenlet "Greenlet<fn:_wrapped_initialize_web_rtc>" at 0x7f27a39b4040: <bound method WebRTCManager._wrapped_initialize_web_rtc of <raiden.network.transport.matrix.rtc.web_rtc.WebRTCManager object at 0x7f27a3be63d0>>(b'\xc0m\xae\xb0\xa9\x91\x81\x11$\xd47FL\xad0\x8e\x)>
            Parent: <Hub '' at 0x7f27adcfef40 epoll default pending=0 ref=20 fileno=4 thread_ident=0x7f27afa81740>
            Running:
              File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 500, in _wrapped_initialize_web_rtc
                self._initialize_web_rtc(address)
              File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 525, in _initialize_web_rtc
                if self._stop_event.wait(timeout=_WEB_RTC_CHANNEL_TIMEOUT):
            Spawned at:
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/runner.py", line 162, in pytest_runtest_call
                item.runtest()
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
                self.ihook.pytest_pyfunc_call(pyfuncitem=self)
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/hooks.py", line 289, in __call__
                return self._hookexec(self, self.get_hookimpls(), kwargs)
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 87, in _hookexec
                return self._inner_hookexec(hook, methods, kwargs)
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/manager.py", line 78, in <lambda>
                self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
                res = hook_impl.function(*args)
              File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
                result = testfunction(**testargs)
              File "/home/konrad/Projects/raiden/raiden/tests/unit/network/rtc/test_web_rtc.py", line 55, in test_basics
                manager1.health_check(address2)
              File "/home/konrad/Projects/raiden/raiden/network/transport/matrix/rtc/web_rtc.py", line 586, in health_check
                self._schedule_new_greenlet(self._wrapped_initialize_web_rtc, partner_address)
              File "/home/konrad/Projects/raiden/raiden/utils/runnable.py", line 89, in _schedule_new_greenlet
                greenlet = Greenlet(func, *args, **kwargs)
---- Thread boundary
<ThreadPoolWorker at 0x7f27a38685c0 thread_ident=0x7f27a1835640 threadpool-hub=<Hub at 0x7f27a395b1c0 thread_ident=0x7f27a3839640>>
 :    Parent: <greenlet.greenlet object at 0x7f27a3868880>
---- Thread boundary
<ThreadPoolWorker at 0x7f27a3868e00 thread_ident=0x7f27a1034640 threadpool-hub=<Hub at 0x7f27a38ec280 thread_ident=0x7f27a3038640>>
 :    Parent: <greenlet.greenlet object at 0x7f27a386eeb0>
---- Thread boundary
<ThreadPoolWorker at 0x7f27a386aa90 thread_ident=0x7f278b7fe640 threadpool-hub=<Hub at 0x7f27a38ec040 thread_ident=0x7f27a2036640>>
 :    Parent: <greenlet.greenlet object at 0x7f27a386ae00>
---- Thread boundary
<ThreadPoolWorker at 0x7f27a386aeb0 thread_ident=0x7f278bfff640 threadpool-hub=<Hub at 0x7f27a38ec340 thread_ident=0x7f27a2837640>>
 :    Parent: <greenlet.greenlet object at 0x7f27a386eca0>
---- Thread boundary
<greenlet.greenlet object at 0x7f27a38463b0>
 :    Parent: None
 +--- <Hub '' at 0x7f27a38ec280 epoll pending=0 ref=0 fileno=28 resolver=<gevent.resolver.thread.Resolver at 0x7f27a3848430 pool=<ThreadPool at 0x7f27a385a3a0 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec280 thread_ident=0x7f27a3038640>>> threadpool=<ThreadPool at 0x7f27a385a3a0 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec280 thread_ident=0x7f27a3038640>> thread_ident=0x7f27a3038640>
            Parent: <greenlet.greenlet object at 0x7f27a38463b0>
---- Thread boundary
<greenlet.greenlet object at 0x7f27a3846720>
 :    Parent: None
 +--- <Hub '' at 0x7f27a38ec340 epoll pending=0 ref=0 fileno=33 resolver=<gevent.resolver.thread.Resolver at 0x7f27a384f880 pool=<ThreadPool at 0x7f27a385a340 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec340 thread_ident=0x7f27a2837640>>> threadpool=<ThreadPool at 0x7f27a385a340 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec340 thread_ident=0x7f27a2837640>> thread_ident=0x7f27a2837640>
            Parent: <greenlet.greenlet object at 0x7f27a3846720>
---- Thread boundary
<greenlet.greenlet object at 0x7f27a3850ca0>
 :    Parent: None
 +--- <Hub '' at 0x7f27a38ec040 epoll pending=0 ref=0 fileno=35 resolver=<gevent.resolver.thread.Resolver at 0x7f27a384fd60 pool=<ThreadPool at 0x7f27a385a400 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec040 thread_ident=0x7f27a2036640>>> threadpool=<ThreadPool at 0x7f27a385a400 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a38ec040 thread_ident=0x7f27a2036640>> thread_ident=0x7f27a2036640>
            Parent: <greenlet.greenlet object at 0x7f27a3850ca0>
---- Thread boundary
<greenlet.greenlet object at 0x7f27a38b8f60>
 :    Parent: None
 +--- <Hub '' at 0x7f27a395b1c0 epoll pending=0 ref=0 fileno=26 resolver=<gevent.resolver.thread.Resolver at 0x7f27a383dcd0 pool=<ThreadPool at 0x7f27a383aee0 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a395b1c0 thread_ident=0x7f27a3839640>>> threadpool=<ThreadPool at 0x7f27a383aee0 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f27a395b1c0 thread_ident=0x7f27a3839640>> thread_ident=0x7f27a3839640>
            Parent: <greenlet.greenlet object at 0x7f27a38b8f60>
Stack for <Task pending name='Task-25' coro=<RTCIceTransport._monitor() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py:343> wait_for=<Future pending cb=[shield.<locals>._outer_done_callback() at /usr/lib/python3.9/asyncio/tasks.py:903, <TaskWakeupMethWrapper object at 0x7f27a38ae460>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py", line 343, in _monitor
    event = await self._connection.get_event()
Stack for <Task pending name='Task-20' coro=<RTCIceTransport._monitor() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py:343> wait_for=<Future pending cb=[shield.<locals>._outer_done_callback() at /usr/lib/python3.9/asyncio/tasks.py:903, <TaskWakeupMethWrapper object at 0x7f27a0767400>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcicetransport.py", line 343, in _monitor
    event = await self._connection.get_event()
Stack for <Task pending name='Task-31' coro=<Connection.query_consent() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aioice/ice.py:936> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f27a3848160>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aioice/ice.py", line 936, in query_consent
    await asyncio.sleep(CONSENT_INTERVAL * (0.8 + 0.4 * random.random()))
Stack for <Task pending name='Task-29' coro=<Connection.query_consent() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aioice/ice.py:936> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f27a38ae820>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aioice/ice.py", line 936, in query_consent
    await asyncio.sleep(CONSENT_INTERVAL * (0.8 + 0.4 * random.random()))
Stack for <Task pending name='Task-35' coro=<RTCDtlsTransport.__run() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcdtlstransport.py:531> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f27a07674f0>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcdtlstransport.py", line 531, in __run
    await self._recv_next()
Stack for <Task pending name='Task-34' coro=<RTCDtlsTransport.__run() running at /home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcdtlstransport.py:531> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f27a3848100>()]>> (most recent call last):
  File "/home/konrad/.virtualenvs/raiden/lib/python3.9/site-packages/aiortc/rtcdtlstransport.py", line 531, in __run
    await self._recv_next()
```

netcriptus commented on raiden/utils/debugging.py:
```diff
@@ -38,6 +45,16 @@ def on_signal(signalnum: Any, stack_frame: Any) -> None:  # pylint: disable=unus
+        asyncio.get_running_loop()
+    except RuntimeError:
+        return
+    tasks = [task for task in asyncio.all_tasks() if task is not asyncio.current_task()]
```
Is it worth it to save `asyncio.current_task()` into a variable and avoid calling the method every time?

istankovic replied:
We should probably also avoid creating the list only to iterate over it -- we can just iterate, e.g.:
```python
for task in asyncio.all_tasks():
    if task is not asyncio.current_task():
        task.print_stack()
```

ulope replied:
Why do we exclude the `current_task()`?

netcriptus reviewed (approved):
One comment, feel free to disregard if you don't agree. LGTM!

konradkonrad committed:
Simpler asyncio debugging

Following PR review suggestions, I simplified the code path for the
asyncio task debugging.
